org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/lookup/IntersectionTypeBinding18.java

```diff
@@ -266,6 +266,17 @@ public char[] qualifiedSourceName() {
 		return qualifiedSourceName.toString().toCharArray();
 	}
 
+	@Override
+	public char[] qualifiedPackageName() {
+		StringBuilder qualifiedPackageName = new StringBuilder(16);
```

Comment:
Is this implementation accurate according to https://docs.oracle.com/javase/specs/jls/se8/jls8.pdf 4.9 ? and also 4.4 ?

Reply:
Hard for me to say. There is no mention of packages, only types. I can ask on the `compiler-dev` mailing list?

Reply:
`qualifiedPackageName()` is never really used in compiler land, so I see no need to consult JLS for this. Any mailing list outside JDT will not know what we use the package name for.

OTOH, I would be surprised if any code in JDT can make sense of an '&' joined sequence of package names.

I haven't debugged the test, but I assume, we only expect one property from this "package name": that it exists (and is equal to itself), am I right?

In other parts of JDT I see lots of callers of `qualifiedPackageName()` for which that &-joined name doesn't make any sense, notably when that is composed with a type name to yield a fully qualified type name.

Would it make sense to actually prevent the call that causes the exception during search?

Alternatively, to just prevent the exception, we could also consider returning a constant synthetic name, like `"<package intersection>"`, which is guaranteed to never occur in source?

Reply:
> In other parts of JDT I see lots of callers of `qualifiedPackageName()` for which that &-joined name doesn't make any sense, notably when that is composed with a type name to yield a fully qualified type name.
> 
> Would it make sense to actually prevent the call that causes the exception during search?

The many callers is why I wouldn't try to avoid the `qualifiedPackageName()` call. We can prevent the auto-completion case here, but in https://bugs.eclipse.org/bugs/show_bug.cgi?id=559168 the problem apparently occurred during a call hierarchy search. Who knows where calling `qualifiedPackageName()` will explode due to `getPackage()` throwing `UnsupportedOperationException`.

> Alternatively, to just prevent the exception, we could also consider returning a constant synthetic name, like `"<package intersection>"`, which is guaranteed to never occur in source?

Initially I wanted to return `CharOperation.NO_CHAR`, but I thought maybe something like `qualifiedSourceName()` makes sense. I'll add a constant as you suggest.

Reply:
> Hard for me to say. There is no mention of packages, only types. I can ask on the `compiler-dev` mailing list?

I think @stephan-herrmann answered this above. But my kind of cryptic question was whether returning concat packages with `&` is correct when reading how type binding is resolved as defined in JLS. But since packages are not used in compiler this will only cause a side effect if we try to use the package name for some type info lookup thinking its a real package name.